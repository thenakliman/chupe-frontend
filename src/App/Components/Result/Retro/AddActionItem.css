.add-action-item-button {
    padding: 0.8rem 1rem;
    color: white;
    cursor: pointer;
    margin: 2rem;
}

.add-action-item-container {
    display: flex;
    flex-direction: column;
    background: #fff;
    padding: 4rem 3rem;
    border: 1px solid #ccc;
    top: 11rem;
    left: 41rem;
    position: fixed;
    width: 35rem;
}

.action-item-due-date {
  width: 10rem;
  margin-bottom: 3rem;
  padding: 0.5rem;
}

.action-item-assigned-to {
  width: 10rem;
  margin-bottom: 1rem;
  padding: 0.5rem;
}

.add-action-item-creation-input {
    padding: 0.8rem 0;
    width: 100%;
    font-size: 1.4rem;
    margin: 2rem 0;
    height: 6rem;
}