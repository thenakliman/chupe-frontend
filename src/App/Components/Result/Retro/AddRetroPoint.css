.add-retro-point-button {
  padding: 0.8rem 1rem;
  color: white;

  cursor: pointer;
  margin: 2rem;
}

.add-retro-point-container {
  display: flex;
  flex-direction: column;
  background: #fff;
  padding: 4rem 3rem;
  border: 1px solid #ccc;
  margin: 7% 32%;
  position: fixed;
}

.add-retro-point-creation-input {
  padding: 0.8rem 0;
  width: 100%;
  font-size: 1.4rem;
  margin: 2rem 0;
}