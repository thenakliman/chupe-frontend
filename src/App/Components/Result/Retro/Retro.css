.retro-container {
    margin: 1rem 18rem;
}

.retro-button-container {
  float: right;
  margin-bottom: 2rem;
}

.retro-row {
  cursor: pointer;
}

.retro-button {
    border-radius: 0.25rem;
    border: 1px solid #28a745;
    background-color: #28a745;
    padding: 0.8rem 2rem;
    float: right;
    margin: 1rem;
    color: white;
    background-image: linear-gradient(-180deg, #34d058 0%, #28a745 90%);
    font-size: 1.15rem;
    cursor: pointer;
}