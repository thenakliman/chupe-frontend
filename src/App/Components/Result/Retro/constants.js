export const DONE_WELL = 'DONE_WELL';
export const ACTION_ITEM = 'ACTION_ITEM';
export const NEED_IMPROVEMENT = 'NEED_IMPROVEMENT';
export const MINIMUM_VOTES = 1;

export const retroState = {
  CREATED: 'CREATED',
  IN_PROGRESS: 'IN_PROGRESS',
  CLOSED: 'CLOSED',
};

export const retroStateToMessageMapping = {
  CREATED: 'Start',
  IN_PROGRESS: 'Close',
};
