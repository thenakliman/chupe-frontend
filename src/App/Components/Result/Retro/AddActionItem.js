import React from 'react';
import propTypes from 'prop-types';

require('./AddActionItem.css');

export class AddActionItem extends React.Component {
    constructor(props) {
      super(props);
      this.state = {description: '', assignedTo: '', deadlineToAct: ''};
      this.handleRetroDescription = this.handleRetroDescription.bind(this);
      this.handleActionItemAssignedToChange = this.handleActionItemAssignedToChange.bind(this);
      this.handleDueDateChange = this.handleDueDateChange.bind(this);
      this.createRetroActionItem = this.createRetroActionItem.bind(this);
    }

    handleRetroDescription(description) {
      this.setState({description: description});
    }

    handleActionItemAssignedToChange(assignedTo) {
      this.setState({assignedTo: assignedTo});
    }

    handleDueDateChange(deadlineToAct) {
      this.setState({deadlineToAct: deadlineToAct});
    }

    createRetroActionItem() {
      this.props.createActionItem(this.state);
      this.setState({description: '', assignedTo: '', deadlineToAct: ''});
    }

    render() {
      return (
      <div className='retro-pop-up-container'>
        <div id='add-action-item-container-id'
             className='add-action-item-container'>
                Description
                <textarea className={'add-action-item-creation-input'}
                       id={'add-action-item-description'}
                       maxLength={2000}
                       onChange={(event) =>
                          this.handleRetroDescription(event.target.value)}
                       value={this.state.description}
                />
                Due Date
                <input className={'action-item-due-date'}
                       id={'action-item-due-date'}
                       onChange={(event) =>
                          this.handleDueDateChange(event.target.value)}
                       value={this.state.deadlineToAct}
                       type={'date'}
                />
                Assigned To
                <select value={this.state.assignedTo}
                        className={'action-item-assigned-to'}
                        onChange={(event) =>
                            this.handleActionItemAssignedToChange(event.target.value)}>
                  <option>Select user</option>
                  {
                    this.props.users.map((user) =>
                      <option key={user.userName}
                              value={user.userName}>
                           {user.userName}
                      </option>
                    )
                  }
                </select>
                <div className='add-action-item-buttons-container'>
                    <button id='add-action-item-button-id'
                      className='add-action-item-button'
                      type='button'
                      onClick={this.createRetroActionItem}
                    >
                      Add Action Item
                    </button>
                    <button id='cancel-add-action-item-button-id'
                      className='add-action-item-button'
                      onClick={this.props.closeAddActionItemPopUp}
                      type='button'
                    >
                      Cancel
                    </button>
                </div>
        </div>
      </div>);
    }
}

AddActionItem.propTypes = {
  closeAddActionItemPopUp: propTypes.func.isRequired,
  createActionItem: propTypes.func.isRequired,
  users: propTypes.array.isRequired,
};
