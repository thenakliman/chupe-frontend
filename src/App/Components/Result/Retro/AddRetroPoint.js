import React from 'react';
import propTypes from 'prop-types';

require('./AddRetroPoint.css');

export class AddRetroPoint extends React.Component {
    constructor(props) {
      super(props);
      this.state = {description: ''};
      this.handleRetroDescription = this.handleRetroDescription.bind(this);
      this.createRetroPoint = this.createRetroPoint.bind(this);
    }

    handleRetroDescription(description) {
      this.setState({description: description});
    }

    createRetroPoint() {
      this.props.createRetroPoint(this.state.description);
      this.setState({description: ''});
    }

    render() {
      return (
      <div className='retro-pop-up-container'>
        <div id='add-retro-point-container-id'
             className='add-retro-point-container'>
                <input className={'add-retro-point-creation-input'}
                       id={'add-retro-point-description'}
                       onChange={(event) =>
                          this.handleRetroDescription(event.target.value)}
                       value={this.state.description}
                />
                <div className='add-retro-point-buttons-container'>
                    <button id='add-retro-point-button-id'
                      className='add-retro-point-button'
                      type='button'
                      onClick={this.createRetroPoint}
                    >
                      Add Retro Point
                    </button>
                    <button id='cancel-add-retro-point-button-id'
                      className='add-retro-point-button'
                      onClick={this.props.closeAddRetroPointPopUp}
                      type='button'
                    >
                      Cancel
                    </button>
                </div>
        </div>
      </div>);
    }
}

AddRetroPoint.propTypes = {
  closeAddRetroPointPopUp: propTypes.func.isRequired,
  createRetroPoint: propTypes.func.isRequired,
};
