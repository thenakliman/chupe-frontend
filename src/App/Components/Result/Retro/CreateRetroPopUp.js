import React from 'react';
import propTypes from 'prop-types';
import {MINIMUM_VOTES} from './constants';

require('./Retros.css');

export class CreateRetroPopUp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {name: '', maximumVote: MINIMUM_VOTES};
    this.handleRetroName = this.handleRetroName.bind(this);
    this.handleMaximumVote = this.handleMaximumVote.bind(this);
    this.createRetro = this.createRetro.bind(this);
  }

  handleRetroName(name) {
    this.setState({name: name});
  }

  handleMaximumVote(maximumVote) {
    this.setState({maximumVote: maximumVote});
  }

  createRetro() {
    this.props.createRetro({
      name: this.state.name,
      maximumVote: this.state.maximumVote,
    });
    this.setState({name: '', maximumVote: MINIMUM_VOTES});
  }

  render() {
    return (
        <div className='retro-pop-up-container'>
          <div id='create-retro-container-id' className='create-retro-container'>
            <div className={'retro-button-container'}>
              <input className={'retro-creation-input'}
                     id={'create-retro-name'}
                     onChange={
                       (event) => this.handleRetroName(event.target.value)}
                     value={this.state.name}
              />
              <input className={'maximum-vote-count'}
                     id={'create-retro-maximum-vote-count'}
                     onChange={
                       (event) => this.handleMaximumVote(event.target.value)}
                     value={this.state.maximumVote}
              />
              <div className='create-retro-buttons-container'>
                <button id='create-retro-button-id'
                        className='create-retro-button'
                        type='button'
                        onClick={this.createRetro}
                >
                  Create Retro
                </button>
                <button id='cancel-create-retro-button-id'
                        className='create-retro-button'
                        onClick={this.props.closeCreatePopUp}
                        type='button'
                >
                  Cancel
                </button>
              </div>
            </div>
          </div>
        </div>);
  }
}

CreateRetroPopUp.propTypes = {
  retros: propTypes.array,
  closeCreatePopUp: propTypes.func.isRequired,
  createRetro: propTypes.func.isRequired,
};
