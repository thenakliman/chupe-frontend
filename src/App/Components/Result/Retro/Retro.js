import React from 'react';
import propTypes from 'prop-types';
import {ACTION_ITEM, DONE_WELL, NEED_IMPROVEMENT, retroStateToMessageMapping} from './constants';
/* eslint-disable */
import {AddRetroPoint} from './AddRetroPoint'
import {AddActionItem} from './AddActionItem'
import {history} from '../../../utils/history'
/* eslint-enable */
require('./Retro.css');

export class Retro extends React.Component {
  constructor(props) {
    super(props);
    this.state = {creatingRetroPoint: false, creatingRetroPointType: ''};
    this.createRetroPoint = this.createRetroPoint.bind(this);
    this.createActionItem = this.createActionItem.bind(this);
    this.changeStatus = this.changeStatus.bind(this);
    this.toggleCreatingRetroPoint = this.toggleCreatingRetroPoint.bind(this);
  }

  toggleCreatingRetroPoint(type) {
    this.setState({
      creatingRetroPoint: !this.state.creatingRetroPoint,
      creatingRetroPointType: type,
    });
  }

  componentWillMount() {
    this.props.getRetroPoints(this.props.match.params.id);
    this.props.getActionItems(this.props.match.params.id);
    this.props.getAllRetros();
    this.props.getUsers();
  }

  createRetroPoint(description) {
    this.props.createRetroPoint({
      description: description,
      type: this.state.creatingRetroPointType,
      retroId: this.props.match.params.id,
    });
    this.toggleCreatingRetroPoint();
    this.setState({creatingRetroPointType: ''});
  }

  createActionItem(actionItem) {
    history.push('/practices-assessment');
    actionItem['retroId'] = this.props.match.params.id;
    this.props.createActionItem(actionItem);
    this.toggleCreatingRetroPoint();
    this.setState({creatingRetroPointType: ''});
  }

  changeStatus(status) {
    this.props.changeStatus(this.props.match.params.id, status);
  }

  renderActionItem() {
    return this.props.actionItems.map(
        (actionItem) => <div key={actionItem.id} id={`action-item-${actionItem.id}`} className={'action-item'}>
          <div className={'action-item-header'}>
            <div className={'action-item-text'} id={`action-item-text-${actionItem.id}`}>
              Assigned To:
            </div>
            <div className={'action-item-assigned'} id={`action-item-${actionItem.id}`}>
              {actionItem.assignedTo}
            </div>
          </div>
          <div className={'retro-point-description'} id={`action-item-description-${actionItem.id}`}>
            {actionItem.description}
          </div>
        </div>
    );
  }

  renderBasedOnFilter(filterFunction) {
    return this.props.retroPoints.map(
        (retroPoint) => filterFunction(retroPoint) &&
            <div key={retroPoint.id} id={`retro-point-${retroPoint.id}`} className={'retro-point'}>
              <div className={'retro-point-header'}>
                <div className={'retro-point-vote-text'} id={`retro-point-vote-text-${retroPoint.id}`}
                     onClick={() => this.props.vote(this.props.match.params.id, retroPoint.id)}>
                  Vote
                </div>
                <div className={'retro-point-votes'} id={`retro-point-votes-${retroPoint.id}`}>
                  {retroPoint.votes}
                </div>
              </div>
              <div className={'retro-point-description'} id={`retro-point-description-${retroPoint.id}`}>
                {retroPoint.description}
              </div>
            </div>
    );
  }

  render() {
    const status = this.props.retros.filter(
        (retro) => retro.id === parseInt(this.props.match.params.id))[0].status;

    const nextStatus = {
      'CREATED': 'IN_PROGRESS',
      'IN_PROGRESS': 'CLOSED',
    };

    return (<div id='retro-points-container-id' className='retro-points-container'>
      {['CREATED', 'IN_PROGRESS'].includes(status) && <button
          id={'change-status-id'}
          className={'change-status-button'}
          onClick={() => this.changeStatus(nextStatus[status])}>
        {retroStateToMessageMapping[status]}
      </button>}
      {this.state.creatingRetroPoint &&
      [DONE_WELL, NEED_IMPROVEMENT].includes(this.state.creatingRetroPointType) &&
      <AddRetroPoint createRetroPoint={(description) => this.createRetroPoint(description)}
                     closeAddRetroPointPopUp={this.toggleCreatingRetroPoint}/>}
      {this.state.creatingRetroPoint &&
      [ACTION_ITEM].includes(this.state.creatingRetroPointType) &&
      <AddActionItem users={this.props.users} createActionItem={this.createActionItem}
                     closeAddActionItemPopUp={this.toggleCreatingRetroPoint}/>}
      <div className='retro-point-header-container'>
        <div className='retro-point-section'>
          Need Improvement
        </div>
        <button
            id='retro-button-for-need-improvement-id'
            className='add-retro-point'
            type='button'
            onClick={() => this.toggleCreatingRetroPoint(NEED_IMPROVEMENT)}>
          Add Point
        </button>
      </div>
      <hr className='horizontal-line'/>
      <div className='retro-point-category'>
        {this.renderBasedOnFilter(
            (retroPoint) => NEED_IMPROVEMENT === retroPoint.type)}
      </div>
      <div className='retro-point-header-container'>
        <div className='retro-point-section'>
          Done Well
        </div>
        <button
            id='retro-button-for-done-well-id'
            className='add-retro-point'
            type='button'
            onClick={() => this.toggleCreatingRetroPoint(DONE_WELL)}>
          Add Point
        </button>
      </div>
      <hr className='horizontal-line'/>
      <div className='retro-point-category'>
        {this.renderBasedOnFilter(
            (retroPoint) => DONE_WELL === retroPoint.type)}
      </div>
      {status === 'IN_PROGRESS' && <div className={'retro-action-item-container'}>
        <div className='retro-point-header-container'>
          <div className='retro-point-section'>Action Items</div>
          <button id='retro-button-for-action-item-id'
                  className='add-retro-point'
                  type='button'
                  onClick={() => this.toggleCreatingRetroPoint(ACTION_ITEM)}
          >
            Add Action Item
          </button>
        </div>
        <hr className='horizontal-line'/>
        <div className='retro-point-category'>
          {this.renderActionItem()}
        </div>
      </div>}
    </div>);
  }
}

Retro.propTypes = {
  actionItems: propTypes.array.isRequired,
  users: propTypes.array.isRequired,
  createActionItem: propTypes.func.isRequired,
  retroPoints: propTypes.array.isRequired,
  getAllRetros: propTypes.func.isRequired,
  vote: propTypes.func.isRequired,
  getRetroPoints: propTypes.func.isRequired,
  createRetroPoint: propTypes.func.isRequired,
  getUsers: propTypes.func.isRequired,
  changeStatus: propTypes.func.isRequired,
};
