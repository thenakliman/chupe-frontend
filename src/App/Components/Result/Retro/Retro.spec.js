/* eslint-disable */
import React from 'react';
import {Retro} from './Retro';
/* eslint-enable */
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

const retros = [{
  name: 'retro - 1',
  id: 10,
  createdBy: 'fake user - 1',
  }, {
  name: 'retro - 2',
  id: 11,
  createdBy: 'fake user - 2',
}];

describe('Show retro component', () => {
  it('should call getAllRetros on component did mount ', () => {
    const getAllRetros = jest.fn();
    shallow(
        <Retro
          retros={retros}
          getAllRetros={getAllRetros}
        />
    );

    expect(getAllRetros).toHaveBeenCalledWith();
  });

  it('should show row for each retro', () => {
    const getAllRetros = jest.fn();
    const wrapper = shallow(
        <Retro
          retros={retros}
          getAllRetros={getAllRetros}
        />
    );

    expect(wrapper.find('#all-retro-table-body-id').children().length).toBe(2);
  });

  describe('should match snapshot', () => {
    const wrapper = shallow(<Retro
       retros={retros}
       getAllRetros={jest.fn()}
    />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
