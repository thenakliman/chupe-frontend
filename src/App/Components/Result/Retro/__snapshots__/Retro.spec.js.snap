// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="retro-container"
  id="retro-container-id"
>
  <div
    className="retro-button-container"
  >
    <button
      className="retro-button"
      id="retro-button-id"
      type="button"
    >
      Create Retro
    </button>
  </div>
  <div>
    <table
      id="all-retros-ordered-list"
    >
      <thead>
        <tr
          className="retro-table-header"
        >
          <th>
             Id 
          </th>
          <th>
             Name 
          </th>
          <th>
             Created By 
          </th>
          <th>
             Maximum Vote 
          </th>
        </tr>
      </thead>
      <tbody
        id="all-retro-table-body-id"
      >
        <tr
          key="10"
        >
          <td>
             
            10
             
          </td>
          <td
            className="retro-row"
            onClick={[Function]}
          >
            retro - 1
          </td>
          <td>
             
            fake user - 1
          </td>
          <td>
             
          </td>
        </tr>
        <tr
          key="11"
        >
          <td>
             
            11
             
          </td>
          <td
            className="retro-row"
            onClick={[Function]}
          >
            retro - 2
          </td>
          <td>
             
            fake user - 2
          </td>
          <td>
             
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
